### PR TITLE
Specify DHCP domain for hostname

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

In 9a8c6db448f200ddd9a06813affab804b183de20, we looked at the hostname
in the metadata service and used '.' as the delimiter to chop off the
dhcp_domain (specified in nova.conf). However administrators need to
better control the dhcp domain better as there may be a '.' in the host
name itself. So let's introduce a config option that we can use and
default it to what nova uses when dhcp_domain is not specified which is
"novalocal"

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
new dhcp-domain parameter to be used for figuring out the hostname of a node
```

### CODE REVIEW

**Comment by dims** [comment]
cc @voelzmo

**Comment by voelzmo** [comment]
Hey @dims,

I'm not sure if this PR goes into the right direction. As proposed in https://github.com/kubernetes/kubernetes/issues/57765#issuecomment-356526418, https://github.com/kubernetes/kubernetes/issues/61774#issuecomment-376779698, and https://github.com/kubernetes/kubernetes/pull/61000#issuecomment-376787212 I think we should not use `metadata.hostname`, but rather `metadata.UUID` instead.

**Comment by dims** [comment]
@voelzmo this PR fixes the problem you mentioned about "what happens if the hostname has a dot". are there other problems/issues that i am not aware of?

**Comment by voelzmo** [comment]
@dims Here's a quick summary
* Before #58502, we had an issue with taking the VM name as nodename, because it can contain arbitrary characters, such as slashes, unicode, etc, which aren't conforming to DNS rfc. See #57765 for the discussion of this issue.
* In this issue, I suggested using one of two things other than the VM name: VM UUID or hostname
  * UUIDS are somewhat defined in format and are strings of numbers separated by dashes
  * hostname is generated by sanitizing the VM name and appending a local domain
* Then #58502 switched to using the hostname, where I asked about the local domain at the end
* This never worked, see #61774, because we were still using the VM name to find a VM given a nodename –  which was no longer the case https://github.com/kubernetes/kubernetes/blob/e3ac40264fa34bac9e4d1c95e56d1d6203554511/pkg/cloudprovider/providers/openstack/openstack.go#L419-L421 
* You fixed this quickly with dropping the local domain. 

Here is the issue with this approach:
* This fix works, if and only if the VM name doesn't need sanitation. If we have a funny VM name such as in my original example:
> E.g. some weird display name like `lol/what?This is weird. Ümlauts rule!` gets a reasonable hostname: 
screen shot 2018-01-10 at 08 52 49
this *doesn't* work, as there's no way to compute the VM name from the sanitized hostname.

In conclusion, the current implementation is working under the exact same conditions as the original implementation which used VM name instead of hostname. It also doesn't work in the case when the VM name contains funny characters, as in my example.

Therefore, I'd suggest to go back to square one and take the UUID instead. This allows you for a much cleaner way to resolve a VM given a hostname (it just returns the same UUID string), and it allows people to assign arbitrary VM names they like, for one reason or another.
The drawback is: your nodenames contain UUIDs.

WDYT?

**Comment by dims** [comment]
@voelzmo so the outstanding problem is that `if hostname from metadata has been sanitized, then there is no way to find the vm`. right? Can we please create a fresh issue? of course the history above is invaluable as well. thanks for that!

**Comment by dims** [comment]
This commit ties up one loose end, so let's get it in. let's work on if uuid is a good idea on another issue/pr. @FengyunPan2 @NickrenREN does that sound good?

**Review by FengyunPan2 — APPROVED** [review]

**Comment by dims** [comment]
since we dont have another issue yet. leaving my notes here - the code responsible for sanitizing host names in nova is:
https://github.com/openstack/nova/blame/master/nova/utils.py#L543-L580

looks like it has not been touched for a few years. which is a good thing for us :)
